### PR TITLE
Added AWS_OS_* macros to detect OS

### PR DESCRIPTION
**Pull request by justinboswell**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/posix/system_info.c [hunk 1]
```diff
@@ -237,11 +237,11 @@ void s_resolve_cmd(char *cmd, size_t len, struct aws_stack_frame_info *frame) {
 #    endif
 
 size_t aws_backtrace(void **frames, size_t num_frames) {
-    return backtrace(frames, num_frames);
+    return backtrace(frames, (int)num_frames);
```
Comment (graebm) [inline source/posix/system_info.c:240]: `(int)aws_max_size(num_frames, INT_MAX)`?

File: source/posix/system_info.c [hunk 2]
```diff
@@ -237,11 +237,11 @@ void s_resolve_cmd(char *cmd, size_t len, struct aws_stack_frame_info *frame) {
...
 }
 
 char **aws_backtrace_symbols(void *const *frames, size_t stack_depth) {
-    return backtrace_symbols(frames, stack_depth);
+    return backtrace_symbols(frames, (int)stack_depth);
```
Comment (graebm) [inline source/posix/system_info.c:244]: `(int)aws_max_size(stack_depth, INT_MAX)`?

File: include/aws/common/platform.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/config.h>
+
+#ifdef _WIN32
+#    define AWS_OS_WINDOWS
+#elif __APPLE__
+#    define AWS_OS_APPLE
```
Comment (graebm) [inline include/aws/common/platform.h:14]: There's other code in aws-c-common ifdefing on `_WIN32 __APPLE__ and __linux__`
Maybe go replace it with these macros?

Also see some in aws-c-io and aws-c-auth...

Comment (justinboswell) [reply include/aws/common/platform.h:14]: I'll do that in a follow-up, this is to unblock other teams, so I need this first.

**Pull request merged**